File: 2022/day4.py
with open('data/day4', 'r') as f:
    data = [[map(int, y.split('-')) for y in x.strip().split(',')] for x in f.readlines()]

def toset(x): return set(range(x[0], x[1]+1))
def part1_condition(x, y): return len(x & y) == len(x) or len(x & y) == len(y)
def compute(data, condition_f):
    count = 0
    for x, y in data:
        if condition_f(toset(x), toset(y)):
            count += 1
    return count

print(compute(data, part1_condition)) # Part1
print(compute(data, lambda x, y: len(x & y) > 0)) # Part2

File: 2023/Day4.scala
import scala.io.Source
import collection.mutable.HashMap
import scala.math.pow

object Day4 extends App {
  val fname = "data/day4"
  val text = Source.fromFile(fname).getLines().toSeq

  case class CardGame(num: Int, winningNums: Set[Int], myNums: Set[Int]){
    def score = pow(2, myNums.intersect(winningNums).size-1).intValue
    def cardCopies = {
      val overlap = myNums.intersect(winningNums).size
      Range(num+1, num+overlap+1).toSet
    }
  }
  def parse(line: String): CardGame = {
    val s"$num: $winningNums | $myNums" = line
    CardGame(num.split(" +").last.toInt, winningNums.strip.split(" +").map(_.toInt).toSet, myNums.strip.split(" +").map(_.toInt).toSet)
  }

  def part1(fname: String) = {
    Source.fromFile(fname).getLines().map(line => {
      parse(line)
    }).map(_.score).sum
  }

  val a = part1(fname)
  println(a)

  def part2(fname: String) = {
    val cardCounts = new HashMap[Int, Int]()
    val games = Source.fromFile(fname).getLines().map(line => {
      parse(line)
    }).map{c =>
      val existingCopies: Int = cardCounts.getOrElse(c.num, 0) + 1
      c.cardCopies.map { y => cardCounts.put(y, cardCounts.getOrElse(y, 0) + existingCopies) }
    }.toSet
    cardCounts.values.sum + games.size
  }

  val b = part2(fname)
  println(b)
}